Explorer/Assets/Scripts/SceneRuntime/Apis/Modules/EngineApi/SDKObservableEvents/SDKObservableEventsEngineApiWrapper.cs

```diff
@@ -0,0 +1,257 @@
+        private readonly List<SDKObservableEvent> sdkObservableEvents = new ();
+        private readonly HashSet<string> sdkObservableEventSubscriptions = new ();
+        private readonly Dictionary<CRDTEntity, string> userIdEntitiesMap = new ();
+        private readonly ISDKObservableEventsEngineApi engineApi;
+        private readonly ISDKMessageBusCommsControllerAPI commsApi;
+        private bool reportedSceneReady;
+
+        public SDKObservableEventsEngineApiWrapper(ISDKObservableEventsEngineApi api, ISDKMessageBusCommsControllerAPI commsApi, IInstancePoolsProvider instancePoolsProvider, ISceneExceptionsHandler exceptionsHandler) : base(api, instancePoolsProvider, exceptionsHandler)
+        {
+            this.engineApi = api;
+            this.commsApi = commsApi;
+        }
+
+        public override void Dispose()
+        {
+            sdkObservableEvents.Clear();
+            sdkObservableEventSubscriptions.Clear();
+            base.Dispose();
+        }
+
+        // Used for SDK Observables + SDK Comms MessageBus
+        [UsedImplicitly]
+        public List<SDKObservableEvent> SendBatch()
+        {
+            if (sdkObservableEventSubscriptions.Count == 0)
+            {
+                engineApi.ClearOutgoingCRDTMessages();
+                return EMPTY_EVENTS_LIST;
+            }
+
+            try
+            {
+                sdkObservableEvents.Clear();
+
+                DetectObservableEventsFromComponents();
+
+                // SDK 'comms' observable for scenes MessageBus
+                DetectSceneMessageBusCommsObservableEvent();
+
+                return sdkObservableEvents;
+            }
+            catch (Exception e)
+            {
+                // Report an uncategorized MANAGED exception (don't propagate it further)
+                exceptionsHandler.OnEngineException(e);
+                return EMPTY_EVENTS_LIST;
+            }
+        }
+
+        private void DetectObservableEventsFromComponents()
+        {
+            // 1. Due to uncertainty in CRDT messages order we have to first check those which populate the userIdEntitiesMap
+            foreach (ProcessedCRDTMessage outgoingCRDTMessage in engineApi.OutgoingCRDTMessages)
+            {
+                CRDTMessage message = outgoingCRDTMessage.message;
+
+                if (message.ComponentId != ComponentID.PLAYER_IDENTITY_DATA || message.Type != CRDTMessageType.PUT_COMPONENT)
+                    continue;
+
+                // onEnterScene + playerConnected observables
+                bool onEnterSceneSubscribed = sdkObservableEventSubscriptions.Contains(SDKObservableEventIds.EnterScene);
+                bool onPlayerConnectedSubscribed = sdkObservableEventSubscriptions.Contains(SDKObservableEventIds.PlayerConnected);
+
+                if (onEnterSceneSubscribed || onPlayerConnectedSubscribed)
+                    playerIdentityDataSerializer.DeserializeInto(playerIdentityData, message.Data.Memory.Span);
```

Comment:
I think it's a weird flow with deserialization after component was serialized (from its managed form).
I propose to change it and integrate it into `OutgoingCRDTMessagesProvider.GetSerializationSyncBlock` as there you have original messages and can get them from there before serialization.

I can provide some help if needed.

Reply:
Uhmm yeah I agreee, I've been trying to tackle this after your comment but I took a look at `OutgoingCRDTMessagesProvider` and in `GetSerializationSyncBlock()` if we deal with those `PendingMessage` we still have to deserialize them using `pendingMessage.Bridge.Serializer.DeserializeInto()` to be able to read the component properties right?